src/data/list/basic.lean

```diff
@@ -1584,6 +1584,25 @@ begin
     exact lt_of_add_lt_add_left' (lt_of_le_of_lt h $ add_lt_add_right (lt_of_not_ge h') _) }
 end
 
+/-- A list of natural numbers with positive sum must have positive length. -/
+-- This could perhaps be generalised, to ordered additive monoids bounded below by zero.
+-- I'm not sure what those are called.
+lemma length_pos_of_sum_pos (L : list ℕ) (h : 0 < L.sum) : 0 < L.length :=
+by { cases L, { cases h, }, { simp, }, }
+
+-- Several lemmas about sum/head/tail for `list ℕ`.
+-- These rely on the fact that `default ℕ = 0`.
+
+lemma head_add_tail_sum (L : list ℕ) : L.head + L.tail.sum = L.sum :=
```

Comment:
This seems true in general

Reply:
No, the problem is that `list.head` requires an inhabited instance, and if the value of default were anything other than `zero`, this would break!  I don't think there's any sensible way to insist that when something is both `[inhabited]` and `[has_zero]`, they are compatible. :-)

Reply:
Maybe `list.head` should require `nonempty` instead? Howcome this breaks if the default is not `zero`? That seems really crazy.

Reply:
Introduce a typeclass (on inhabited has_zero types) saying that default is zero?

Reply:
The point is that `list.head` always returns something --- it returns the default value when the list is empty.

Reply:
It would be true in general if we added the hypothesis `0 < L.length`.

Reply:
Ok, how do you feel about proving the general case under that assumption, and also giving this hacky case with a `'` (or `_nat`) attached to the name?

Reply:
I think I'd prefer to prove the `(L.nth 0).get_or_else 0 + L.tail.sum = L.sum` version. Either way is pretty gross.